src/resolve/resolved-dns-dnssec.c

```diff
@@ -760,24 +846,62 @@ int dnssec_verify_rrset(
 
                 /* Hash the source of synthesis. If this is a wildcard, then prefix it with the *. label */
                 if (wildcard)
-                        gcry_md_write(md, (uint8_t[]) { 1, '*'}, 2);
-                gcry_md_write(md, wire_format_name, r);
+                        fwrite((uint8_t[]) { 1, '*'}, sizeof(uint8_t), 2, f);
+                fwrite(wire_format_name, 1, r, f);
 
-                md_add_uint16(md, rr->key->type);
-                md_add_uint16(md, rr->key->class);
-                md_add_uint32(md, rrsig->rrsig.original_ttl);
+                fwrite_uint16(rr->key->type, f);
+                fwrite_uint16(rr->key->class, f);
+                fwrite_uint32(rrsig->rrsig.original_ttl, f);
 
                 l = DNS_RESOURCE_RECORD_RDATA_SIZE(rr);
                 assert(l <= 0xFFFF);
 
-                md_add_uint16(md, (uint16_t) l);
-                gcry_md_write(md, DNS_RESOURCE_RECORD_RDATA(rr), l);
+                fwrite_uint16((uint16_t) l, f);
+                fwrite(DNS_RESOURCE_RECORD_RDATA(rr), 1, l, f);
         }
 
-        hash = gcry_md_read(md, 0);
-        if (!hash) {
-                r = -EIO;
+        if (ferror(f)) {
+                fclose(f);
+                free(sig_data);
+                return -ENOMEM;
+        } else
+                fclose(f);
+
+        switch (rrsig->rrsig.algorithm) {
```

Comment:
The control flow within dnssec_verify_rrset is somewhat complicated. I would appreciate if someone verify that it's correct.